coalib/output/printers/LogPrinter.py

```diff
@@ -1,54 +1,194 @@
+            cb=color_begin,
+            ce=color_end,
+            levelname=record.levelname)
+        return super(ColoredFormatter, self).format(record, *args, **kwargs)
+
+
 class LogPrinter:
     """
     The LogPrinter class allows to print log messages to an underlying Printer.
 
     This class is an adapter, means you can create a LogPrinter from every
     existing Printer instance.
     """
+    _loggers = {}
+
+    @staticmethod
+    def _get_logger(printer, logger_name, time_fmt=None):
+        """
+        Returns existing or newly created and configured logger.
+        :param logger_name: name of the logger to be returned.
+        :param printer:     printer to initialize logger with if it doesn't
+                            exist, yet.
+        :param time_fmt:    format to be used for time output. No time is
+                            pinter if None.
+        """
+        if not LogPrinter._loggers.get(logger_name):
+            if time_fmt:
+                format_str = '%(levelname)s[%(asctime)s] %(message)s'
+            else:
+                format_str = '%(levelname)s %(message)s'
+
+            use_color = isinstance(printer, ColorPrinter)
+            formatter = ColoredFormatter(format_str,
+                                         time_fmt,
+                                         use_color=use_color)
+
+            handler = PrinterHandler(printer)
+            handler.setFormatter(formatter)
+
+            logger = logging.getLogger(logger_name)
+            logger.addHandler(handler)
+
+            LogPrinter._loggers[logger_name] = logger
+        return LogPrinter._loggers[logger_name]
 
     def __init__(self,
-                 printer,
+                 printer=None,
                  log_level=LOG_LEVEL.INFO,
-                 timestamp_format="%X"):
+                 timestamp_format="%X",
+                 of_type=True):
         """
-        Creates a new log printer from an existing Printer.
+        Creates a new log printer from an existing Printer or returns existing
+        one if it already exist.
+        Note: log_level changes on every logger instantiation.
 
         :param printer:          The underlying Printer where log messages
                                  shall be written to. If you inherit from
-                                 LogPrinter, set it to self.
+                                 Printer, set it to self.
         :param log_level:        The minimum log level, everything below will
                                  not be logged.
         :param timestamp_format: The format string for the
                                  datetime.today().strftime(format) method.
+        :param of_type:          If False returns new logger object every time.
+                                 Else may return logger which uses the same
+                                 printer type or will use printer to initialize
+                                 new one.
```

Comment:
Feels weird, kind of singleton but not really, smells a bit to me. Is it needed or can we just drop the singleton  aspect?